src/main/java/com/turn/ttorrent/common/Torrent.java

```diff
@@ -418,6 +418,12 @@ public void save(OutputStream output) throws IOException {
 		return md.digest();
 	}
 
+	public static byte[] hash(ByteBuffer data) throws NoSuchAlgorithmException {
+		MessageDigest md = MessageDigest.getInstance("SHA-1");
+		md.update(data);
```

Comment:
The buffer being passed is the data read from the "output" file (`this._read(...)`), which is a direct buffer, i.e. is not array backed, no ? The piecesHashes one is array backed.

I couldn't understand the rest of your answer, sorry. Upon reading the source of the MessageDigest I can see this  optimization is there, reusing the byte array internally `tempArray`, but here md is created and discarded just as well, I think the solution lies more in the lines of #184, of having a "resource store" where you pick it up, use it and relinquish ownership...